cpp/arcticdb/processing/clause.cpp

```diff
@@ -658,4 +658,107 @@ std::string DateRangeClause::to_string() const {
     return fmt::format("DATE RANGE {} - {}", start_, end_);
 }
 
+
+
+TopKClause::TopKClause(std::vector<float_t> query_vector, uint8_t k) :
+            query_vector(std::move(query_vector)),
+            k(k) {
+        clause_info_.modifies_output_descriptor_ = true;
+        clause_info_.follows_original_columns_order_ = false;
+    } // Do we want/need to move?
+
+Composite<ProcessingSegment> TopKClause::process(std::shared_ptr<Store> store,
+                                                 Composite<ProcessingSegment> &&p) const {
+    auto procs = std::move(p);
+    std::set<NamedColumnSortedByDistance> top_k;
+    double_t furthest_distance = std::numeric_limits<double_t>::infinity();
+    // We expect a vector in each column.
+    // top_k is updated as we read more vectors. We want to efficiently remove
+    // the 'worst' (least similar) vector as we go along to ensure we have the
+    // top k, not the top k+1.
+    auto lp = 2;
+    // Let's pretend that only the lp-norms exist. We'll use p=2 for now as a default.
+
+    procs.broadcast(
+            [&, lp] (ProcessingSegment &proc) { // todo: cleanup references
+                std::for_each(
+                        proc.data().begin(),
+                        proc.data().end(),
+                        [&] (pipelines::SliceAndKey s) {
+                            std::vector<std::shared_ptr<Column>> columns = s.segment(store).columns();
+                            for (auto column : folly::enumerate(columns)) {
```

Comment:
I find:
`for (auto&& [idx, column] : folly::enumerate(columns))`
more readable personally